<a href="https://colab.research.google.com/github/JuanMaLR/bedu-equipo4-dataScience-P-FP/blob/second/FinalProject.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Proyecto Final Fase 2 Módulo 2 BEDU

##Procesamiento de datos con Python - Equipo 4

###Integrantes: 
- Juan Manuel Ledesma Rangel
- Hannia Melendres Samaniego
- Alejandro Cortéz Sánchez
- Martín Otamendi Torres
- Rodrigo Alvarado Gómez


---
##A. Identificación de un problema - Postwork 1

###A.1 Problema:
> Conocimiento respecto al derecho a protección y/o asistencia consular como mexicano migrante dentro de territorio estadounidense.

###A.2 Justificación del problema:
> La migración de ciudadanos mexicanos a Estados Unidos es un tema que está en la conciencia poblacional. No es novedad el flujo de migrantes hacia este país en busca de mejores y nuevas oportunidades. Sin embargo, poca es la educación respecto a los derechos que se tienen como mexicano en territorio estadounidense, entre ellos el derecho a la protección y/o asistencia consular. Por tanto, y con ayuda de los datos proporcionados por la Secretaría de Relaciones Exteriores (SRE), se desea analizar los casos de protección y/o asistencia consular atendidos en la Red Diplomática Consular de México en Estados Unidos (para el ámbito migratorio), con la finalidad de darlos a conocer.

###A.3 Investigación preliminar:
> De acuerdo a la CNDH [1], los mexicanos migrantes tienen derechos reconocidos en la Constitución Política de los Estados Unidos Mexicanos. Entre ellos, el derecho a la asistencia consular y a la protección. Es trabajo de la Embajada de México hacer regir estos derechos en todos los países con relaciones diplomáticas. En Estados Unidos de América, los trámites de protección en materia de migración son los derivados de casos de repatriación y detención migratoria [2]. A su vez, los datos que serán analizados, aportados por la SRE [3], indican como casos de ámbito migracional a las remociones, permisos de internación, menores bajo la ley TVPRA (menores que cruzan la frontera sin acompañamiento adulto [4]), usos de documentos falsos, víctimas de fraude migratorio, y la orientación general sobre consultas migratorias y/o trámites de residencia.

###A.4 Soluciones anteriores al problema:
> Si bien no se encontraron investigaciones puntuales respecto al dataset en concreto, sí se han realizado esfuerzos por hacer llegar la información a más personas. La CNDH, por ejemplo, ha publicado un archivo pdf [5] de fácil lectura hablando sobre los derechos de migrantes.

---
##B. Planteamiento de Preguntas - Postwork 2

El tema de la migración trae, sin dudas, muchísimas preguntas de la mano. Sin embargo, para los registros de trámites realizados en los consulados mexicanos dentro EUA, nosotros nos cuestionamos:

1.   ¿Cuál es el trámite más consultado?

2.   ¿Cuál es el trámite menos consultado?

3.   ¿Cuál estado de la República Mexicana tiene más flujo migratorio?

4.   ¿Cuál es el porcentaje de menores de edad bajo la ley TVPRA (menores que migraron sin acompañamiento adulto)?

5.   Si deseo realizar un trámite en época de pandemia y evitar dentro de lo posible una aglomeración de personas, ¿a cuál consulado debería asistir de acuerdo al número total de casos atendidos entre 2019 y 2022?

6.   ¿Cómo ha cambiado el flujo migratorio por estado conforme a la llegada de la pandemia?



---
## C. Colección de Datos/Uso de APIs  - Postwork 3 y 6

Para el análisis, utilizaremos el dataset (en formato .csv)

>"Migratorio - Casos de protección y/o asistencia consular atendidos por la RDCM en el mundo"

aportado por la Secretaría de Relaciones Exteriores (SRE) desde la página de datos.gob.mx [3].

Para realizar el análisis exploratorio de datos (AED), importamos la librería "pandas" y convertimos el archivo .csv a dataframe desde su página de origen:

In [1]:
import pandas as pd

#Archivo CSV original recabado de la página del gobierno
#Fue necesario agregar "encoding= 'latin-1'" para poder leerlo

df = pd.read_csv('https://portales.sre.gob.mx/transparencia/datos_abiertos/DGPME/Migratorio_CasodeproteccionconsularatendidosporlaRDCMx.csv',
                 sep=',', encoding= 'latin-1')

---
##D. Análisis exploratorio de datos - Postwork 4

Hacemos una previsualización de los datos, para definir si será necesario hacer limpieza:

In [2]:
#Hacemos una previsualización de los datos

df.head(15)

Casos de proteccion  y/o asistencia consular atendidos en la RDCM en Estados Unidos - ámbito migratorio  \
0                                                 NaN                                                        
1                                                 AÑO                                                        
2                                              ÁMBITO                                                        
3                                          MIGRATORIO                                                        
4                                                 NaN                                                        
5                                                 NaN                                                        
6                                                 NaN                                                        
7                                                 NaN                                                        
8                                                 NaN                                                        
9                                                 NaN                                                        
10                                                NaN                                                        
11                                                NaN                                                        
12                                                NaN                                                        
13                                                NaN                                                        
14                                                NaN                                                        

                                           Unnamed: 1 Unnamed: 2 Unnamed: 3  \
0                                                 NaN        NaN        NaN   
1                                                 NaN       2010        NaN   
2                                 RUBRO DE PROTECCIÓN      DESC.     HOMBRE   
3     INCIDENTES, MÚLTIPLES INVOLUCRADOS (MIGRATORIO)        NaN        NaN   
4                      MENOR PROCESADO BAJO LEY TVPRA          1        373   
5   ORIENTACIóN GENERAL SOBRE CONSULTAS MIGRATORIA...          2      1,734   
6                   OTROS CASOS DEL áMBITO MIGRATORIO         10      4,423   
7    PERMISOS DE INTERNACIóN POR RAZONES HUMANITARIAS          6      1,143   
8         PROCESOS JUDICIALES MIGRATORIOS DE REMOCIóN         16     10,455   
9       PROCESOS JUDICIALES MIGRATORIOS DE RESIDENCIA        NaN        339   
10            REMOCIóN EXPEDITA EN PUERTOS DE ENTRADA         41     21,203   
11   REPATRIACIóN DE ENFERMOS EN PROCESOS DE REMOCIóN        NaN        178   
12  REPATRIACIóN DE ENFERMOS, PERSONAS ACCIDENTADA...          1        727   
13  REPATRIACIóN DE MENORES (POR ESTANCIA O INGRES...         21     10,225   
14  REPATRIACIóN DE MENORES INFRACTORES (NO POR FA...        NaN        505   

   Unnamed: 4 Unnamed: 5 Unnamed: 6 Unnamed: 7 Unnamed: 8 Unnamed: 9  ...  \
0         NaN        NaN        NaN        NaN        NaN        NaN  ...   
1         NaN       2011        NaN        NaN       2012        NaN  ...   
2       MUJER      DESC.     HOMBRE      MUJER      DESC.     HOMBRE  ...   
3         NaN        NaN        NaN        NaN        NaN        NaN  ...   
4         135        NaN        375        112        NaN        357  ...   
5       1,472          6      1,844      1,885          4      1,687  ...   
6       1,026          6      8,073      1,129         11      8,545  ...   
7       1,352        NaN      1,096      1,403          1        902  ...   
8       1,174         18     18,514      1,413         26     49,479  ...   
9         413          1        775        304          1        260  ...   
10      3,897          2     32,763      1,428          2     32,491  ...   
11         25         53        158         14       

##Análisis incial:
Como podemos observar, todas nuestras columnas aparecen como "Unnamed" salvo la primera que dice: "Casos de proteccion y/o asistencia consular atendidos en la RDCM en Estados Unidos - ámbito migratorio". 

También se puede observar que la fila cero tiene puros NaNs y que la fila 1 y 2 parecen ser encabezados indicando el año de los casos atendido y el género de la persona cuyo caso fue atendido respectivamente. 

Por otro lado, se puede observar que a partir de la columna "Unnamed: 42" se tienen puros NaNs en los número de casos atendidos. 

Por lo anterior este dataset es ideal para poner en práctica los conocimientos adquiridos en este curso. 

A fin de poder entender mejor el dataset será necesario observar otros parámetros de utilidad cómo sus dimensiones, columnas, tipos de dato, etc.

In [3]:
#Para ver las dimensiones
df.shape

(219, 52)

In [4]:
#Para ver las columnas que lo integran
df.columns

Index(['Casos de proteccion  y/o asistencia consular atendidos en la RDCM en Estados Unidos - ámbito migratorio',
       'Unnamed: 1', 'Unnamed: 2', 'Unnamed: 3', 'Unnamed: 4', 'Unnamed: 5',
       'Unnamed: 6', 'Unnamed: 7', 'Unnamed: 8', 'Unnamed: 9', 'Unnamed: 10',
       'Unnamed: 11', 'Unnamed: 12', 'Unnamed: 13', 'Unnamed: 14',
       'Unnamed: 15', 'Unnamed: 16', 'Unnamed: 17', 'Unnamed: 18',
       'Unnamed: 19', 'Unnamed: 20', 'Unnamed: 21', 'Unnamed: 22',
       'Unnamed: 23', 'Unnamed: 24', 'Unnamed: 25', 'Unnamed: 26',
       'Unnamed: 27', 'Unnamed: 28', 'Unnamed: 29', 'Unnamed: 30',
       'Unnamed: 31', 'Unnamed: 32', 'Unnamed: 33', 'Unnamed: 34',
       'Unnamed: 35', 'Unnamed: 36', 'Unnamed: 37', 'Unnamed: 38',
       'Unnamed: 39', 'Unnamed: 40', 'Unnamed: 41', 'Unnamed: 42',
       'Unnamed: 43', 'Unnamed: 44', 'Unnamed: 45', 'Unnamed: 46',
       'Unnamed: 47', 'Unnamed: 48', 'Unnamed: 49', 'Unnamed: 50',
       'Unnamed: 51'],
      dtype='object')

In [5]:
#Para ver los tipos de dato por columna
df.dtypes

Casos de proteccion  y/o asistencia consular atendidos en la RDCM en Estados Unidos - ámbito migratorio     object
Unnamed: 1                                                                                                  object
Unnamed: 2                                                                                                  object
Unnamed: 3                                                                                                  object
Unnamed: 4                                                                                                  object
Unnamed: 5                                                                                                  object
Unnamed: 6                                                                                                  object
Unnamed: 7                                                                                                  object
Unnamed: 8                                                                      

De lo anterior podemos observar que tenemos dos tipos de dato en nuestro dataframe, object y float64. Si observamos con mayor detalle veremos que la razón por la que tenemos object es porque las primeras dos filas que tendrían que ser encabezados son contenido y al detectar texto pandas convierte toda la columna a objeto. 

Por último, para simplificar todo lo que vimos con anterioridad y para también visualizar los NaNs en nuestro dataframe utilizaremos info: 

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 219 entries, 0 to 218
Data columns (total 52 columns):
 #   Column                                                                                                   Non-Null Count  Dtype  
---  ------                                                                                                   --------------  -----  
 0   Casos de proteccion  y/o asistencia consular atendidos en la RDCM en Estados Unidos - ámbito migratorio  17 non-null     object 
 1   Unnamed: 1                                                                                               204 non-null    object 
 2   Unnamed: 2                                                                                               170 non-null    object 
 3   Unnamed: 3                                                                                               179 non-null    object 
 4   Unnamed: 4                                                                            

De esta manera observamos que todas nuestras columnas contienen al menos 1 elemento NaN, ya que tenemos 219 filas y en el "Non-Null Count" de nuestra función info vemos que el número más cercano a 219 es 204. Por lo tanto aquí hay 15 NaNs. El mismo razonamiento puede utilizarse con el resto de las columnas. 

###D.1 Primer problema:
Este dataset, a pesar de estar en formato .csv, fue tratado como un archivo .xlsx. Esto lo podemos notar si abrimos el archivo en Excel, ya que encontramos varias tablas en una sola hoja de cálculo, además de comentarios en múltiples celdas y espacios en blanco, como se muestra en la siguiente imagen:

![Estructura en crudo del csv.png](https://github.com/JuanMaLR/bedu-equipo4-dataScience-P-FP/blob/second/images/raw_struc.png?raw=true)
(Las líneas rojas delimitan 3 de las 8 tablas presentes en la misma hoja de cálculo, únicamente para ejemplificar)

Por lo anterior decidimos no continuar con el AED de este dataframe y propusimos separar en tablas individuales el csv inicial. 

Por motivos del scope del proyecto (que es sólo casos en EUA) sólo fueron de nuestro interés 4 de las 8 tablas encontradas en el dataframe incial. 

Cada tabla fue tratada como un archivo .csv indiviual, y se subió a nuestro repositorio de GitHub dentro de la carpeta "raw_data". A continuación sus lecturas en solitario:

In [124]:
#De las 8 tablas, se leen únicamente aquellas cuya información es relevante para el análisis del caso de EUA (4)
#Las convertimos a dataframe con pandas

casos = pd.read_csv('https://raw.githubusercontent.com/JuanMaLR/bedu-equipo4-dataScience-P-FP/second/raw_data/casos.csv', sep=',', encoding= 'latin-1')
oficinas = pd.read_csv('https://raw.githubusercontent.com/JuanMaLR/bedu-equipo4-dataScience-P-FP/second/raw_data/oficinas.csv', sep=',', encoding= 'latin-1')
edo_origen = pd.read_csv('https://raw.githubusercontent.com/JuanMaLR/bedu-equipo4-dataScience-P-FP/second/raw_data/edo_origen.csv', sep=',', encoding= 'latin-1')
rango_edad = pd.read_csv('https://raw.githubusercontent.com/JuanMaLR/bedu-equipo4-dataScience-P-FP/second/raw_data/rango_edad.csv', sep=',', encoding= 'latin-1')

Verificamos que se hayan cargado correctamente utilizando head:

In [9]:
#Visualizando casos.csv

casos.head()

AÑO                                         Unnamed: 1   2010  \
0      ÁMBITO                                RUBRO DE PROTECCIÓN  DESC.   
1  MIGRATORIO    INCIDENTES, MÚLTIPLES INVOLUCRADOS (MIGRATORIO)    NaN   
2         NaN                     MENOR PROCESADO BAJO LEY TVPRA      1   
3         NaN  ORIENTACIóN GENERAL SOBRE CONSULTAS MIGRATORIA...      2   
4         NaN                  OTROS CASOS DEL áMBITO MIGRATORIO     10   

  Unnamed: 3 Unnamed: 4   2011 Unnamed: 6 Unnamed: 7   2012 Unnamed: 9  ...  \
0     HOMBRE      MUJER  DESC.     HOMBRE      MUJER  DESC.     HOMBRE  ...   
1        NaN        NaN    NaN        NaN        NaN    NaN        NaN  ...   
2        373        135    NaN        375        112    NaN        357  ...   
3      1,734      1,472      6      1,844      1,885      4      1,687  ...   
4      4,423      1,026      6      8,073      1,129     11      8,545  ...   

  Unnamed: 31   2020 Unnamed: 33 Unnamed: 34   2021 Unnamed: 36 Unnamed: 37  \
0       MUJER  DESC.      HOMBRE       MUJER  DESC.      HOMBRE       MUJER   
1         NaN    NaN         NaN         NaN    NaN         NaN         NaN   
2         312      1         230         128      2         726         526   
3       7,948      7       2,584       2,451      4       2,229       1,115   
4       6,519    142      20,194       3,248      3      12,009       2,985   

  * 2022 Unnamed: 39 Unnamed: 40  
0  DESC.      HOMBRE       MUJER  
1    NaN          23           1  
2      2         467         359  
3      2         693         852  
4      1       8,004       1,730  

[5 rows x 41 columns]

In [10]:
#Visualizando oficinas.csv

oficinas.head()

AÑO      Unnamed: 1   2019 Unnamed: 3 Unnamed: 4   2020 Unnamed: 6  \
0      ÁMBITO  REPRESENTACIÓN  DESC.     HOMBRE      MUJER  DESC.     HOMBRE   
1  MIGRATORIO     ALBUQUERQUE      1        260        173      2        162   
2         NaN         ATLANTA      0        660        285      2        287   
3         NaN          AUSTIN      1        690        429      1        171   
4         NaN           BOISE      0         82         63      3         62   

  Unnamed: 7   2021 Unnamed: 9  ... Unnamed: 31 Unnamed: 32 Unnamed: 33  \
0      MUJER  DESC.     HOMBRE  ...         NaN         NaN         NaN   
1         59      1         47  ...         NaN         NaN         NaN   
2        164    NaN        146  ...         NaN         NaN         NaN   
3        144    NaN         91  ...         NaN         NaN         NaN   
4         47    NaN         16  ...         NaN         NaN         NaN   

  Unnamed: 34  Unnamed: 35  Unnamed: 36  Unnamed: 37  Unnamed: 38  \
0         NaN          NaN          NaN          NaN          NaN   
1         NaN          NaN          NaN          NaN          NaN   
2         NaN          NaN          NaN          NaN          NaN   
3         NaN          NaN          NaN          NaN          NaN   
4         NaN          NaN          NaN          NaN          NaN   

   Unnamed: 39  Unnamed: 40  
0          NaN          NaN  
1          NaN          NaN  
2          NaN          NaN  
3          NaN          NaN  
4          NaN          NaN  

[5 rows x 41 columns]

In [11]:
#Visualizando edo_origen.csv

edo_origen.head()

AÑO  Unnamed: 1  Unnamed: 2   2019 Unnamed: 4 Unnamed: 5  \
0          EDO. ORIGEN         NaN         NaN  DESC.     HOMBRE      MUJER   
1       AGUASCALIENTES         NaN         NaN      1        815        184   
2      BAJA CALIFORNIA         NaN         NaN      1        894        316   
3  BAJA CALIFORNIA SUR         NaN         NaN      0         41         23   
4             CAMPECHE         NaN         NaN      1        255         46   

    2020 Unnamed: 7 Unnamed: 8   2021  ... Unnamed: 31 Unnamed: 32  \
0  DESC.     HOMBRE      MUJER  DESC.  ...         NaN         NaN   
1      0        785         89    NaN  ...         NaN         NaN   
2      0        943        215      1  ...         NaN         NaN   
3      0         31          8    NaN  ...         NaN         NaN   
4      1        266         21    NaN  ...         NaN         NaN   

  Unnamed: 33 Unnamed: 34 Unnamed: 35  Unnamed: 36  Unnamed: 37  Unnamed: 38  \
0         NaN         NaN         NaN          NaN          NaN          NaN   
1         NaN         NaN         NaN          NaN          NaN          NaN   
2         NaN         NaN         NaN          NaN          NaN          NaN   
3         NaN         NaN         NaN          NaN          NaN          NaN   
4         NaN         NaN         NaN          NaN          NaN          NaN   

   Unnamed: 39  Unnamed: 40  
0          NaN          NaN  
1          NaN          NaN  
2          NaN          NaN  
3          NaN          NaN  
4          NaN          NaN  

[5 rows x 41 columns]

In [12]:
#Visualizando rango_edad.csv

rango_edad.head()

AÑO    2019 Unnamed: 2 Unnamed: 3    2020 Unnamed: 5  \
0    Grupo de edad   DESC.     HOMBRE      MUJER   DESC.     HOMBRE   
1  De 00 a 17 años      50      7,375      1,985      15      5,569   
2  De 18 a 25 años      78     20,381      4,514      57     18,793   
3  De 26 a 35 años     143     32,117      5,733      82     27,261   
4  De 36 a 45 años      93     24,630      5,599      40     19,468   

  Unnamed: 6   2021 Unnamed: 8 Unnamed: 9  ... Unnamed: 31 Unnamed: 32  \
0      MUJER  DESC.     HOMBRE      MUJER  ...         NaN         NaN   
1      1,326      1     11,737      2,128  ...         NaN         NaN   
2      2,827     11     10,538      1,903  ...         NaN         NaN   
3      3,425      7     11,122      2,021  ...         NaN         NaN   
4      2,513      7      8,729      1,570  ...         NaN         NaN   

  Unnamed: 33  Unnamed: 34  Unnamed: 35  Unnamed: 36  Unnamed: 37  \
0         NaN          NaN          NaN          NaN          NaN   
1         NaN          NaN          NaN          NaN          NaN   
2         NaN          NaN          NaN          NaN          NaN   
3         NaN          NaN          NaN          NaN          NaN   
4         NaN          NaN          NaN          NaN          NaN   

   Unnamed: 38  Unnamed: 39  Unnamed: 40  
0          NaN          NaN          NaN  
1          NaN          NaN          NaN  
2          NaN          NaN          NaN  
3          NaN          NaN          NaN  
4          NaN          NaN          NaN  

[5 rows x 41 columns]

A simple vista resaltan las áreas de mejora de los dataframes que se mencionaron en un inicio: 


*   Filas de encabezados incorrectas
*   Procesamiento de NaNs
*   Tipos de dato incorrectos
*   Columnas innecesarias para nuestro planteamiento inicial


Sin embargo, observamos el resumen de cada uno de ellos para conocerlos más a fondo con ayuda de la función "info":

In [14]:
#Resumen de los dataframes - casos
print("Dataframe de casos")
print(casos.info())

Dataframe de casos
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17 entries, 0 to 16
Data columns (total 41 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   AÑO          2 non-null      object
 1   Unnamed: 1   17 non-null     object
 2   2010         12 non-null     object
 3   Unnamed: 3   16 non-null     object
 4   Unnamed: 4   16 non-null     object
 5   2011         11 non-null     object
 6   Unnamed: 6   16 non-null     object
 7   Unnamed: 7   16 non-null     object
 8   2012         10 non-null     object
 9   Unnamed: 9   16 non-null     object
 10  Unnamed: 10  16 non-null     object
 11  2013         8 non-null      object
 12  Unnamed: 12  16 non-null     object
 13  Unnamed: 13  16 non-null     object
 14  2014         9 non-null      object
 15  Unnamed: 15  16 non-null     object
 16  Unnamed: 16  16 non-null     object
 17  2015         12 non-null     object
 18  Unnamed: 18  16 non-null     object
 19  Unnamed: 19 

In [15]:
#Resumen de los dataframes - oficinas
print("Dataframe de oficinas")
print(oficinas.info())

Dataframe de oficinas
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 52 entries, 0 to 51
Data columns (total 41 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   AÑO          2 non-null      object 
 1   Unnamed: 1   52 non-null     object 
 2   2019         52 non-null     object 
 3   Unnamed: 3   52 non-null     object 
 4   Unnamed: 4   52 non-null     object 
 5   2020         52 non-null     object 
 6   Unnamed: 6   52 non-null     object 
 7   Unnamed: 7   52 non-null     object 
 8   2021         24 non-null     object 
 9   Unnamed: 9   52 non-null     object 
 10  Unnamed: 10  52 non-null     object 
 11  * 2022       14 non-null     object 
 12  Unnamed: 12  52 non-null     object 
 13  Unnamed: 13  52 non-null     object 
 14  Unnamed: 14  0 non-null      float64
 15  Unnamed: 15  0 non-null      float64
 16  Unnamed: 16  0 non-null      float64
 17  Unnamed: 17  0 non-null      float64
 18  Unnamed: 18  0 non-null      f

In [16]:
#Resumen de los dataframes - estado de orígen
print("Dataframe de estado de orígen")
print(edo_origen.info())

Dataframe de estado de orígen
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 52 entries, 0 to 51
Data columns (total 41 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   AÑO          36 non-null     object 
 1   Unnamed: 1   0 non-null      float64
 2   Unnamed: 2   0 non-null      float64
 3   2019         36 non-null     object 
 4   Unnamed: 4   36 non-null     object 
 5   Unnamed: 5   36 non-null     object 
 6   2020         36 non-null     object 
 7   Unnamed: 7   36 non-null     object 
 8   Unnamed: 8   36 non-null     object 
 9   2021         15 non-null     object 
 10  Unnamed: 10  36 non-null     object 
 11  Unnamed: 11  36 non-null     object 
 12  * 2022       15 non-null     object 
 13  Unnamed: 13  36 non-null     object 
 14  Unnamed: 14  36 non-null     object 
 15  Unnamed: 15  0 non-null      float64
 16  Unnamed: 16  0 non-null      float64
 17  Unnamed: 17  0 non-null      float64
 18  Unnamed: 18  0 non-nul

In [17]:
#Resumen de los dataframes - rango de edad
print("Dataframe de rango de edad")
print(rango_edad.info())

Dataframe de rango de edad
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 52 entries, 0 to 51
Data columns (total 41 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   AÑO          11 non-null     object 
 1   2019         11 non-null     object 
 2   Unnamed: 2   11 non-null     object 
 3   Unnamed: 3   11 non-null     object 
 4   2020         11 non-null     object 
 5   Unnamed: 5   11 non-null     object 
 6   Unnamed: 6   11 non-null     object 
 7   2021         11 non-null     object 
 8   Unnamed: 8   11 non-null     object 
 9   Unnamed: 9   11 non-null     object 
 10  * 2022       11 non-null     object 
 11  Unnamed: 11  11 non-null     object 
 12  Unnamed: 12  11 non-null     object 
 13  Unnamed: 13  0 non-null      float64
 14  Unnamed: 14  0 non-null      float64
 15  Unnamed: 15  0 non-null      float64
 16  Unnamed: 16  0 non-null      float64
 17  Unnamed: 17  0 non-null      float64
 18  Unnamed: 18  0 non-null  

Está función nos arroja información sobre el tipo de datos, el nombre de las columnas, y la cantidad de registros y de no-nulls de un dataframe. Consideramos que esta función es suficiente para conocer la estructura de nuestro dataset, pero las dimensiones del mismo no son tan visibles. Por lo que utilizamos también la función "shape":

In [ ]:
#Dimensiones
print(casos.shape)
print(oficinas.shape)
print(edo_origen.shape)
print(rango_edad.shape)

(17, 41)
(52, 41)
(52, 41)
(52, 41)


A pesar de haber realizado la separación del archivo .csv original en cuatro distintos, los datos siguen estando sucios en demasía. Contienen bastantes datos NotaNumber (NaN), los encabezados no están definidos, la mayoría de tipos de datos son objetos, y hay filas y columnas innecesarias.

Así, procedemos a la limpieza de los datos.

##E. Limpieza del DataFrame - Postwork 5



Para comenzar arreglaremos las filas de encabezado, ya que nos sería imposible realizar cualquier análisis sobre los datos teniendo estas filas dentro de nuestros datos. 

Si recurrimos a los archivos desde una hoja de cálculo, notamos que esto se debe a que en realidad tienen dos encabezados, y todos siguen el mismo patrón:

![Patrón de encabezado.png](https://github.com/JuanMaLR/bedu-equipo4-dataScience-P-FP/blob/second/images/pattern.png?raw=true)

Decidimos unir ambas filas para simplificar el dataframe colocando una d para DESC., h para HOMBRE y m para MUJER, seguido del año en que se realizó el registro. Por ejemplo: h_2010.

Con el fin de no realizarlo "a mano" (y porque es un detalle que se repite en todos los dataframes), creamos una función y la reciclaremos. 

Sin embargo, antes de hacer esto será necesario que los 4 dataframes tengan la misma estructura para que nuestra función pueda aplicarse sin problema. 

Si recordamos, en el punto anterior (D. AED) cuando visualizamos el head de nuestros 4 dataframes pudimos notar que los dataframes de casos y oficinas tienen una primera columna con las filas 0 y 1 con las palabras "ámbito" y "migratorio" respectivamente, pero el resto de la columna tenía puros NaNs. 

Esta columna no nos sirve ya no aporta información por lo tanto la eliminaremos. 

In [23]:
#Para visualizar el dataframe de casos
casos.head(10)

AÑO                                         Unnamed: 1   2010  \
0      ÁMBITO                                RUBRO DE PROTECCIÓN  DESC.   
1  MIGRATORIO    INCIDENTES, MÚLTIPLES INVOLUCRADOS (MIGRATORIO)    NaN   
2         NaN                     MENOR PROCESADO BAJO LEY TVPRA      1   
3         NaN  ORIENTACIóN GENERAL SOBRE CONSULTAS MIGRATORIA...      2   
4         NaN                  OTROS CASOS DEL áMBITO MIGRATORIO     10   
5         NaN   PERMISOS DE INTERNACIóN POR RAZONES HUMANITARIAS      6   
6         NaN        PROCESOS JUDICIALES MIGRATORIOS DE REMOCIóN     16   
7         NaN      PROCESOS JUDICIALES MIGRATORIOS DE RESIDENCIA    NaN   
8         NaN            REMOCIóN EXPEDITA EN PUERTOS DE ENTRADA     41   
9         NaN   REPATRIACIóN DE ENFERMOS EN PROCESOS DE REMOCIóN    NaN   

  Unnamed: 3 Unnamed: 4   2011 Unnamed: 6 Unnamed: 7   2012 Unnamed: 9  ...  \
0     HOMBRE      MUJER  DESC.     HOMBRE      MUJER  DESC.     HOMBRE  ...   
1        NaN        NaN    NaN        NaN        NaN    NaN        NaN  ...   
2        373        135    NaN        375        112    NaN        357  ...   
3      1,734      1,472      6      1,844      1,885      4      1,687  ...   
4      4,423      1,026      6      8,073      1,129     11      8,545  ...   
5      1,143      1,352    NaN      1,096      1,403      1        902  ...   
6     10,455      1,174     18     18,514      1,413     26     49,479  ...   
7        339        413      1        775        304      1        260  ...   
8     21,203      3,897      2     32,763      1,428      2     32,491  ...   
9        178         25     53        158         14    NaN        156  ...   

  Unnamed: 31   2020 Unnamed: 33 Unnamed: 34   2021 Unnamed: 36 Unnamed: 37  \
0       MUJER  DESC.      HOMBRE       MUJER  DESC.      HOMBRE       MUJER   
1         NaN    NaN         NaN         NaN    NaN         NaN         NaN   
2         312      1         230         128      2         726         526   
3       7,948      7       2,584       2,451      4       2,229       1,115   
4       6,519    142      20,194       3,248      3      12,009       2,985   
5       1,045    NaN         510         556    NaN         445         476   
6       5,339     64      48,940       4,564     12      14,908       1,560   
7         325      4         580         231    NaN         734         296   
8         695      5       2,136         535      2       2,286         825   
9           8      1         171          10    NaN          25           2   

  * 2022 Unnamed: 39 Unnamed: 40  
0  DESC.      HOMBRE       MUJER  
1    NaN          23           1  
2      2         467         359  
3      2         693         852  
4      1       8,004       1,730  
5    NaN         126         199  
6      9       8,722       1,006  
7    NaN       1,028          83  
8      3         564         258  
9    NaN          12           2  

[10 rows x 41 columns]

In [27]:
#Para visualizar el dataframe de oficinas
oficinas.head(10)

AÑO      Unnamed: 1   2019 Unnamed: 3 Unnamed: 4   2020 Unnamed: 6  \
0      ÁMBITO  REPRESENTACIÓN  DESC.     HOMBRE      MUJER  DESC.     HOMBRE   
1  MIGRATORIO     ALBUQUERQUE      1        260        173      2        162   
2         NaN         ATLANTA      0        660        285      2        287   
3         NaN          AUSTIN      1        690        429      1        171   
4         NaN           BOISE      0         82         63      3         62   
5         NaN          BOSTON      0         61         16      0         16   
6         NaN     BROWNSVILLE      1      4,899        484      2      4,624   
7         NaN        CALEXICO     15      6,242      1,302      1      6,059   
8         NaN         CHICAGO      1      3,078        781      2      1,586   
9         NaN          DALLAS      4      3,085      1,347     10      5,651   

  Unnamed: 7   2021 Unnamed: 9  ... Unnamed: 31 Unnamed: 32 Unnamed: 33  \
0      MUJER  DESC.     HOMBRE  ...         NaN         NaN         NaN   
1         59      1         47  ...         NaN         NaN         NaN   
2        164    NaN        146  ...         NaN         NaN         NaN   
3        144    NaN         91  ...         NaN         NaN         NaN   
4         47    NaN         16  ...         NaN         NaN         NaN   
5         12    NaN         15  ...         NaN         NaN         NaN   
6        323      1      1,279  ...         NaN         NaN         NaN   
7        491      2      2,026  ...         NaN         NaN         NaN   
8        472      2      1,111  ...         NaN         NaN         NaN   
9        521      1      2,934  ...         NaN         NaN         NaN   

  Unnamed: 34  Unnamed: 35  Unnamed: 36  Unnamed: 37  Unnamed: 38  \
0         NaN          NaN          NaN          NaN          NaN   
1         NaN          NaN          NaN          NaN          NaN   
2         NaN          NaN          NaN          NaN          NaN   
3         NaN          NaN          NaN          NaN          NaN   
4         NaN          NaN          NaN          NaN          NaN   
5         NaN          NaN          NaN          NaN          NaN   
6         NaN          NaN          NaN          NaN          NaN   
7         NaN          NaN          NaN          NaN          NaN   
8         NaN          NaN          NaN          NaN          NaN   
9         NaN          NaN          NaN          NaN          NaN   

   Unnamed: 39  Unnamed: 40  
0          NaN          NaN  
1          NaN          NaN  
2          NaN          NaN  
3          NaN          NaN  
4          NaN          NaN  
5          NaN          NaN  
6          NaN          NaN  
7          NaN          NaN  
8          NaN          NaN  
9          NaN          NaN  

[10 rows x 41 columns]

In [125]:
#Con la visualización anterior se puede apreciar que la columa "Año" es la que debe ser eliminada de ambos dataframes
#Por lo tanto:
casos.drop(columns='AÑO', inplace=True)
casos.head(10)

Unnamed: 1   2010 Unnamed: 3  \
0                                RUBRO DE PROTECCIÓN  DESC.     HOMBRE   
1    INCIDENTES, MÚLTIPLES INVOLUCRADOS (MIGRATORIO)    NaN        NaN   
2                     MENOR PROCESADO BAJO LEY TVPRA      1        373   
3  ORIENTACIóN GENERAL SOBRE CONSULTAS MIGRATORIA...      2      1,734   
4                  OTROS CASOS DEL áMBITO MIGRATORIO     10      4,423   
5   PERMISOS DE INTERNACIóN POR RAZONES HUMANITARIAS      6      1,143   
6        PROCESOS JUDICIALES MIGRATORIOS DE REMOCIóN     16     10,455   
7      PROCESOS JUDICIALES MIGRATORIOS DE RESIDENCIA    NaN        339   
8            REMOCIóN EXPEDITA EN PUERTOS DE ENTRADA     41     21,203   
9   REPATRIACIóN DE ENFERMOS EN PROCESOS DE REMOCIóN    NaN        178   

  Unnamed: 4   2011 Unnamed: 6 Unnamed: 7   2012 Unnamed: 9 Unnamed: 10  ...  \
0      MUJER  DESC.     HOMBRE      MUJER  DESC.     HOMBRE       MUJER  ...   
1        NaN    NaN        NaN        NaN    NaN        NaN         NaN  ...   
2        135    NaN        375        112    NaN        357         194  ...   
3      1,472      6      1,844      1,885      4      1,687       1,789  ...   
4      1,026      6      8,073      1,129     11      8,545       6,359  ...   
5      1,352    NaN      1,096      1,403      1        902       1,084  ...   
6      1,174     18     18,514      1,413     26     49,479       1,172  ...   
7        413      1        775        304      1        260         212  ...   
8      3,897      2     32,763      1,428      2     32,491          94  ...   
9         25     53        158         14    NaN        156          23  ...   

  Unnamed: 31   2020 Unnamed: 33 Unnamed: 34   2021 Unnamed: 36 Unnamed: 37  \
0       MUJER  DESC.      HOMBRE       MUJER  DESC.      HOMBRE       MUJER   
1         NaN    NaN         NaN         NaN    NaN         NaN         NaN   
2         312      1         230         128      2         726         526   
3       7,948      7       2,584       2,451      4       2,229       1,115   
4       6,519    142      20,194       3,248      3      12,009       2,985   
5       1,045    NaN         510         556    NaN         445         476   
6       5,339     64      48,940       4,564     12      14,908       1,560   
7         325      4         580         231    NaN         734         296   
8         695      5       2,136         535      2       2,286         825   
9           8      1         171          10    NaN          25           2   

  * 2022 Unnamed: 39 Unnamed: 40  
0  DESC.      HOMBRE       MUJER  
1    NaN          23           1  
2      2         467         359  
3      2         693         852  
4      1       8,004       1,730  
5    NaN         126         199  
6      9       8,722       1,006  
7    NaN       1,028          83  
8      3         564         258  
9    NaN          12           2  

[10 rows x 40 columns]

In [126]:
oficinas.drop(columns='AÑO', inplace=True)
oficinas.head(10)

Unnamed: 1   2019 Unnamed: 3 Unnamed: 4   2020 Unnamed: 6 Unnamed: 7  \
0  REPRESENTACIÓN  DESC.     HOMBRE      MUJER  DESC.     HOMBRE      MUJER   
1     ALBUQUERQUE      1        260        173      2        162         59   
2         ATLANTA      0        660        285      2        287        164   
3          AUSTIN      1        690        429      1        171        144   
4           BOISE      0         82         63      3         62         47   
5          BOSTON      0         61         16      0         16         12   
6     BROWNSVILLE      1      4,899        484      2      4,624        323   
7        CALEXICO     15      6,242      1,302      1      6,059        491   
8         CHICAGO      1      3,078        781      2      1,586        472   
9          DALLAS      4      3,085      1,347     10      5,651        521   

    2021 Unnamed: 9 Unnamed: 10  ... Unnamed: 31 Unnamed: 32 Unnamed: 33  \
0  DESC.     HOMBRE       MUJER  ...         NaN         NaN         NaN   
1      1         47          13  ...         NaN         NaN         NaN   
2    NaN        146          78  ...         NaN         NaN         NaN   
3    NaN         91          75  ...         NaN         NaN         NaN   
4    NaN         16          25  ...         NaN         NaN         NaN   
5    NaN         15           4  ...         NaN         NaN         NaN   
6      1      1,279         152  ...         NaN         NaN         NaN   
7      2      2,026         262  ...         NaN         NaN         NaN   
8      2      1,111         601  ...         NaN         NaN         NaN   
9      1      2,934         277  ...         NaN         NaN         NaN   

   Unnamed: 34  Unnamed: 35  Unnamed: 36  Unnamed: 37  Unnamed: 38  \
0          NaN          NaN          NaN          NaN          NaN   
1          NaN          NaN          NaN          NaN          NaN   
2          NaN          NaN          NaN          NaN          NaN   
3          NaN          NaN          NaN          NaN          NaN   
4          NaN          NaN          NaN          NaN          NaN   
5          NaN          NaN          NaN          NaN          NaN   
6          NaN          NaN          NaN          NaN          NaN   
7          NaN          NaN          NaN          NaN          NaN   
8          NaN          NaN          NaN          NaN          NaN   
9          NaN          NaN          NaN          NaN          NaN   

   Unnamed: 39  Unnamed: 40  
0          NaN          NaN  
1          NaN          NaN  
2          NaN          NaN  
3          NaN          NaN  
4          NaN          NaN  
5          NaN          NaN  
6          NaN          NaN  
7          NaN          NaN  
8          NaN          NaN  
9          NaN          NaN  

[10 rows x 40 columns]

Ahora, si prestamos atención al dataset de edo_origen veremos que luego de la primer columna tiene dos columnas vacías ("Unnamed: 1" y "Unnamed: 2"). Por lo tanto es necesario eliminarlas. 

In [88]:
#Observamos las columans de "Unnamed: 1" y "Unnamed: 2" que contienen puros NaNs
edo_origen.head(10)

AÑO  Unnamed: 1  Unnamed: 2   2019 Unnamed: 4 Unnamed: 5  \
0          EDO. ORIGEN         NaN         NaN  DESC.     HOMBRE      MUJER   
1       AGUASCALIENTES         NaN         NaN      1        815        184   
2      BAJA CALIFORNIA         NaN         NaN      1        894        316   
3  BAJA CALIFORNIA SUR         NaN         NaN      0         41         23   
4             CAMPECHE         NaN         NaN      1        255         46   
5              CHIAPAS         NaN         NaN     67      4,437      1,159   
6            CHIHUAHUA         NaN         NaN      9      2,353        875   
7     CIUDAD DE MÉXICO         NaN         NaN     16      2,519      1,072   
8             COAHUILA         NaN         NaN      3      1,137        574   
9               COLIMA         NaN         NaN      5        303        115   

    2020 Unnamed: 7 Unnamed: 8   2021  ... Unnamed: 31 Unnamed: 32  \
0  DESC.     HOMBRE      MUJER  DESC.  ...         NaN         NaN   
1      0        785         89    NaN  ...         NaN         NaN   
2      0        943        215      1  ...         NaN         NaN   
3      0         31          8    NaN  ...         NaN         NaN   
4      1        266         21    NaN  ...         NaN         NaN   
5     21      4,413        673      2  ...         NaN         NaN   
6      0      1,549        343      2  ...         NaN         NaN   
7      6      1,762        567    NaN  ...         NaN         NaN   
8      0        566        157    NaN  ...         NaN         NaN   
9      2        261         50    NaN  ...         NaN         NaN   

  Unnamed: 33 Unnamed: 34 Unnamed: 35  Unnamed: 36  Unnamed: 37  Unnamed: 38  \
0         NaN         NaN         NaN          NaN          NaN          NaN   
1         NaN         NaN         NaN          NaN          NaN          NaN   
2         NaN         NaN         NaN          NaN          NaN          NaN   
3         NaN         NaN         NaN          NaN          NaN          NaN   
4         NaN         NaN         NaN          NaN          NaN          NaN   
5         NaN         NaN         NaN          NaN          NaN          NaN   
6         NaN         NaN         NaN          NaN          NaN          NaN   
7         NaN         NaN         NaN          NaN          NaN          NaN   
8         NaN         NaN         NaN          NaN          NaN          NaN   
9         NaN         NaN         NaN          NaN          NaN          NaN   

   Unnamed: 39  Unnamed: 40  
0          NaN          NaN  
1          NaN          NaN  
2          NaN          NaN  
3          NaN          NaN  
4          NaN          NaN  
5          NaN          NaN  
6          NaN          NaN  
7          NaN          NaN  
8          NaN          NaN  
9          NaN          NaN  

[10 rows x 41 columns]

In [127]:
edo_origen.drop(columns=['Unnamed: 1', 'Unnamed: 2'], inplace=True)
edo_origen.head(10)

AÑO   2019 Unnamed: 4 Unnamed: 5   2020 Unnamed: 7  \
0          EDO. ORIGEN  DESC.     HOMBRE      MUJER  DESC.     HOMBRE   
1       AGUASCALIENTES      1        815        184      0        785   
2      BAJA CALIFORNIA      1        894        316      0        943   
3  BAJA CALIFORNIA SUR      0         41         23      0         31   
4             CAMPECHE      1        255         46      1        266   
5              CHIAPAS     67      4,437      1,159     21      4,413   
6            CHIHUAHUA      9      2,353        875      0      1,549   
7     CIUDAD DE MÉXICO     16      2,519      1,072      6      1,762   
8             COAHUILA      3      1,137        574      0        566   
9               COLIMA      5        303        115      2        261   

  Unnamed: 8   2021 Unnamed: 10 Unnamed: 11  ... Unnamed: 31 Unnamed: 32  \
0      MUJER  DESC.      HOMBRE       MUJER  ...         NaN         NaN   
1         89    NaN         429          65  ...         NaN         NaN   
2        215      1         610         155  ...         NaN         NaN   
3          8    NaN          22           9  ...         NaN         NaN   
4         21    NaN         162          17  ...         NaN         NaN   
5        673      2       3,649         551  ...         NaN         NaN   
6        343      2       1,459         279  ...         NaN         NaN   
7        567    NaN         909         438  ...         NaN         NaN   
8        157    NaN         500         109  ...         NaN         NaN   
9         50    NaN         136          43  ...         NaN         NaN   

  Unnamed: 33  Unnamed: 34  Unnamed: 35  Unnamed: 36  Unnamed: 37  \
0         NaN          NaN          NaN          NaN          NaN   
1         NaN          NaN          NaN          NaN          NaN   
2         NaN          NaN          NaN          NaN          NaN   
3         NaN          NaN          NaN          NaN          NaN   
4         NaN          NaN          NaN          NaN          NaN   
5         NaN          NaN          NaN          NaN          NaN   
6         NaN          NaN          NaN          NaN          NaN   
7         NaN          NaN          NaN          NaN          NaN   
8         NaN          NaN          NaN          NaN          NaN   
9         NaN          NaN          NaN          NaN          NaN   

   Unnamed: 38  Unnamed: 39  Unnamed: 40  
0          NaN          NaN          NaN  
1          NaN          NaN          NaN  
2          NaN          NaN          NaN  
3          NaN          NaN          NaN  
4          NaN          NaN          NaN  
5          NaN          NaN          NaN  
6          NaN          NaN          NaN  
7          NaN          NaN          NaN  
8          NaN          NaN          NaN  
9          NaN          NaN          NaN  

[10 rows x 39 columns]

Dentro del mismo dataset de estado origen vamos ahora a observar las columnas de la derecha. 

In [93]:
edo_origen.loc[:, 'Unnamed: 11':'Unnamed: 40'].head(10)

Unnamed: 11 * 2022 Unnamed: 13 Unnamed: 14  Unnamed: 15  Unnamed: 16  \
0       MUJER  DESC.      HOMBRE       MUJER          NaN          NaN   
1          65    NaN         228          28          NaN          NaN   
2         155      1         298          80          NaN          NaN   
3           9    NaN          20          11          NaN          NaN   
4          17    NaN          66          20          NaN          NaN   
5         551      3       2,498         445          NaN          NaN   
6         279    NaN         835         131          NaN          NaN   
7         438      1         504         238          NaN          NaN   
8         109    NaN         328          56          NaN          NaN   
9          43    NaN          63          24          NaN          NaN   

   Unnamed: 17  Unnamed: 18  Unnamed: 19  Unnamed: 20  ...  Unnamed: 31  \
0          NaN          NaN          NaN          NaN  ...          NaN   
1          NaN          NaN          NaN          NaN  ...          NaN   
2          NaN          NaN          NaN          NaN  ...          NaN   
3          NaN          NaN          NaN          NaN  ...          NaN   
4          NaN          NaN          NaN          NaN  ...          NaN   
5          NaN          NaN          NaN          NaN  ...          NaN   
6          NaN          NaN          NaN          NaN  ...          NaN   
7          NaN          NaN          NaN          NaN  ...          NaN   
8          NaN          NaN          NaN          NaN  ...          NaN   
9          NaN          NaN          NaN          NaN  ...          NaN   

   Unnamed: 32  Unnamed: 33  Unnamed: 34  Unnamed: 35  Unnamed: 36  \
0          NaN          NaN          NaN          NaN          NaN   
1          NaN          NaN          NaN          NaN          NaN   
2          NaN          NaN          NaN          NaN          NaN   
3          NaN          NaN          NaN          NaN          NaN   
4          NaN          NaN          NaN          NaN          NaN   
5          NaN          NaN          NaN          NaN          NaN   
6          NaN          NaN          NaN          NaN          NaN   
7          NaN          NaN          NaN          NaN          NaN   
8          NaN          NaN          NaN          NaN          NaN   
9          NaN          NaN          NaN          NaN          NaN   

   Unnamed: 37  Unnamed: 38  Unnamed: 39  Unnamed: 40  
0          NaN          NaN          NaN          NaN  
1          NaN          NaN          NaN          NaN  
2          NaN          NaN          NaN          NaN  
3          NaN          NaN          NaN          NaN  
4          NaN          NaN          NaN          NaN  
5          NaN          NaN          NaN          NaN  
6          NaN          NaN          NaN          NaN  
7          NaN          NaN          NaN          NaN  
8          NaN          NaN          NaN          NaN  
9          NaN          NaN          NaN          NaN  

[10 rows x 30 columns]

Como se puede observar, a partir de la columna "Unnamed: 15" hasta la columna "Unnamed: 40" se tienen puros NaNs en los datos. Por lo tanto podemos eliminar estas columnas. 

In [128]:
columnas_edo_origen_a_eliminar = edo_origen.loc[:, 'Unnamed: 15':'Unnamed: 40'].columns
edo_origen.drop(columns=columnas_edo_origen_a_eliminar, inplace=True)
edo_origen.head(10)

AÑO   2019 Unnamed: 4 Unnamed: 5   2020 Unnamed: 7  \
0          EDO. ORIGEN  DESC.     HOMBRE      MUJER  DESC.     HOMBRE   
1       AGUASCALIENTES      1        815        184      0        785   
2      BAJA CALIFORNIA      1        894        316      0        943   
3  BAJA CALIFORNIA SUR      0         41         23      0         31   
4             CAMPECHE      1        255         46      1        266   
5              CHIAPAS     67      4,437      1,159     21      4,413   
6            CHIHUAHUA      9      2,353        875      0      1,549   
7     CIUDAD DE MÉXICO     16      2,519      1,072      6      1,762   
8             COAHUILA      3      1,137        574      0        566   
9               COLIMA      5        303        115      2        261   

  Unnamed: 8   2021 Unnamed: 10 Unnamed: 11 * 2022 Unnamed: 13 Unnamed: 14  
0      MUJER  DESC.      HOMBRE       MUJER  DESC.      HOMBRE       MUJER  
1         89    NaN         429          65    NaN         228          28  
2        215      1         610         155      1         298          80  
3          8    NaN          22           9    NaN          20          11  
4         21    NaN         162          17    NaN          66          20  
5        673      2       3,649         551      3       2,498         445  
6        343      2       1,459         279    NaN         835         131  
7        567    NaN         909         438      1         504         238  
8        157    NaN         500         109    NaN         328          56  
9         50    NaN         136          43    NaN          63          24

En el caso del dataset de rango_edad sucede algo similar que con edo_origen. Las columnas a la derecha de "Unnamed: 12" están vacías. Por lo que es necesario eliminarlas. 

In [98]:
rango_edad.loc[:, 'Unnamed: 11':'Unnamed: 40'].head(10)

Unnamed: 11 Unnamed: 12  Unnamed: 13  Unnamed: 14  Unnamed: 15  Unnamed: 16  \
0      HOMBRE       MUJER          NaN          NaN          NaN          NaN   
1       7,288       1,325          NaN          NaN          NaN          NaN   
2       4,541         995          NaN          NaN          NaN          NaN   
3       6,208       1,154          NaN          NaN          NaN          NaN   
4       4,986         828          NaN          NaN          NaN          NaN   
5       2,125         588          NaN          NaN          NaN          NaN   
6         684         222          NaN          NaN          NaN          NaN   
7         119          87          NaN          NaN          NaN          NaN   
8          32          36          NaN          NaN          NaN          NaN   
9       1,875         325          NaN          NaN          NaN          NaN   

   Unnamed: 17  Unnamed: 18  Unnamed: 19  Unnamed: 20  ...  Unnamed: 31  \
0          NaN          NaN          NaN          NaN  ...          NaN   
1          NaN          NaN          NaN          NaN  ...          NaN   
2          NaN          NaN          NaN          NaN  ...          NaN   
3          NaN          NaN          NaN          NaN  ...          NaN   
4          NaN          NaN          NaN          NaN  ...          NaN   
5          NaN          NaN          NaN          NaN  ...          NaN   
6          NaN          NaN          NaN          NaN  ...          NaN   
7          NaN          NaN          NaN          NaN  ...          NaN   
8          NaN          NaN          NaN          NaN  ...          NaN   
9          NaN          NaN          NaN          NaN  ...          NaN   

   Unnamed: 32  Unnamed: 33  Unnamed: 34  Unnamed: 35  Unnamed: 36  \
0          NaN          NaN          NaN          NaN          NaN   
1          NaN          NaN          NaN          NaN          NaN   
2          NaN          NaN          NaN          NaN          NaN   
3          NaN          NaN          NaN          NaN          NaN   
4          NaN          NaN          NaN          NaN          NaN   
5          NaN          NaN          NaN          NaN          NaN   
6          NaN          NaN          NaN          NaN          NaN   
7          NaN          NaN          NaN          NaN          NaN   
8          NaN          NaN          NaN          NaN          NaN   
9          NaN          NaN          NaN          NaN          NaN   

   Unnamed: 37  Unnamed: 38  Unnamed: 39  Unnamed: 40  
0          NaN          NaN          NaN          NaN  
1          NaN          NaN          NaN          NaN  
2          NaN          NaN          NaN          NaN  
3          NaN          NaN          NaN          NaN  
4          NaN          NaN          NaN          NaN  
5          NaN          NaN          NaN          NaN  
6          NaN          NaN          NaN          NaN  
7          NaN          NaN          NaN          NaN  
8          NaN          NaN          NaN          NaN  
9          NaN          NaN          NaN          NaN  

[10 rows x 30 columns]

In [129]:
columnas_rango_edad_a_eliminar = rango_edad.loc[:, 'Unnamed: 13':'Unnamed: 40'].columns
rango_edad.drop(columns=columnas_rango_edad_a_eliminar, inplace=True)
rango_edad.head(10)

AÑO    2019 Unnamed: 2 Unnamed: 3    2020 Unnamed: 5  \
0      Grupo de edad   DESC.     HOMBRE      MUJER   DESC.     HOMBRE   
1    De 00 a 17 años      50      7,375      1,985      15      5,569   
2    De 18 a 25 años      78     20,381      4,514      57     18,793   
3    De 26 a 35 años     143     32,117      5,733      82     27,261   
4    De 36 a 45 años      93     24,630      5,599      40     19,468   
5    De 46 a 55 años      48     10,254      3,492      17      7,845   
6    De 56 a 65 años      12      2,646      1,240       2      1,773   
7    De 66 a 75 años       9        597        369       0        385   
8     Más de 75 años       5        191        188       0         91   
9  Se desconoce Edad      15      2,146        595      31      1,864   

  Unnamed: 6   2021 Unnamed: 8 Unnamed: 9 * 2022 Unnamed: 11 Unnamed: 12  
0      MUJER  DESC.     HOMBRE      MUJER  DESC.      HOMBRE       MUJER  
1      1,326      1     11,737      2,128      8       7,288       1,325  
2      2,827     11     10,538      1,903      5       4,541         995  
3      3,425      7     11,122      2,021      2       6,208       1,154  
4      2,513      7      8,729      1,570      5       4,986         828  
5      1,555      2      4,143      1,052      4       2,125         588  
6        532      1      1,044        400      1         684         222  
7        197      0        225        150      0         119          87  
8         89      0         76         81      0          32          36  
9        418     10      1,494        376      2       1,875         325

De igual manera en el caso del dataset de oficinas sucede algo similar. Las columnas a la derecha de "Unnamed: 13" están vacías. Por lo que es necesario eliminarlas. 

In [115]:
oficinas.loc[:, 'Unnamed: 12':'Unnamed: 38'].head(10)

Unnamed: 12 Unnamed: 13  Unnamed: 14  Unnamed: 15  Unnamed: 16  Unnamed: 17  \
0      HOMBRE       MUJER          NaN          NaN          NaN          NaN   
1          11           6          NaN          NaN          NaN          NaN   
2          75          55          NaN          NaN          NaN          NaN   
3         129          98          NaN          NaN          NaN          NaN   
4          38          17          NaN          NaN          NaN          NaN   
5          11           3          NaN          NaN          NaN          NaN   
6       1,936         202          NaN          NaN          NaN          NaN   
7         409          98          NaN          NaN          NaN          NaN   
8         430         142          NaN          NaN          NaN          NaN   
9       1,095          41          NaN          NaN          NaN          NaN   

   Unnamed: 18  Unnamed: 19  Unnamed: 20  Unnamed: 21  ...  Unnamed: 29  \
0          NaN          NaN          NaN          NaN  ...          NaN   
1          NaN          NaN          NaN          NaN  ...          NaN   
2          NaN          NaN          NaN          NaN  ...          NaN   
3          NaN          NaN          NaN          NaN  ...          NaN   
4          NaN          NaN          NaN          NaN  ...          NaN   
5          NaN          NaN          NaN          NaN  ...          NaN   
6          NaN          NaN          NaN          NaN  ...          NaN   
7          NaN          NaN          NaN          NaN  ...          NaN   
8          NaN          NaN          NaN          NaN  ...          NaN   
9          NaN          NaN          NaN          NaN  ...          NaN   

   Unnamed: 30  Unnamed: 31  Unnamed: 32  Unnamed: 33  Unnamed: 34  \
0          NaN          NaN          NaN          NaN          NaN   
1          NaN          NaN          NaN          NaN          NaN   
2          NaN          NaN          NaN          NaN          NaN   
3          NaN          NaN          NaN          NaN          NaN   
4          NaN          NaN          NaN          NaN          NaN   
5          NaN          NaN          NaN          NaN          NaN   
6          NaN          NaN          NaN          NaN          NaN   
7          NaN          NaN          NaN          NaN          NaN   
8          NaN          NaN          NaN          NaN          NaN   
9          NaN          NaN          NaN          NaN          NaN   

   Unnamed: 35  Unnamed: 36  Unnamed: 37  Unnamed: 38  
0          NaN          NaN          NaN          NaN  
1          NaN          NaN          NaN          NaN  
2          NaN          NaN          NaN          NaN  
3          NaN          NaN          NaN          NaN  
4          NaN          NaN          NaN          NaN  
5          NaN          NaN          NaN          NaN  
6          NaN          NaN          NaN          NaN  
7          NaN          NaN          NaN          NaN  
8          NaN          NaN          NaN          NaN  
9          NaN          NaN          NaN          NaN  

[10 rows x 27 columns]

In [130]:
columnas_oficinas_a_eliminar = oficinas.loc[:, 'Unnamed: 14':'Unnamed: 40'].columns
oficinas.drop(columns=columnas_oficinas_a_eliminar, inplace=True)
oficinas.head(10)

Unnamed: 1   2019 Unnamed: 3 Unnamed: 4   2020 Unnamed: 6 Unnamed: 7  \
0  REPRESENTACIÓN  DESC.     HOMBRE      MUJER  DESC.     HOMBRE      MUJER   
1     ALBUQUERQUE      1        260        173      2        162         59   
2         ATLANTA      0        660        285      2        287        164   
3          AUSTIN      1        690        429      1        171        144   
4           BOISE      0         82         63      3         62         47   
5          BOSTON      0         61         16      0         16         12   
6     BROWNSVILLE      1      4,899        484      2      4,624        323   
7        CALEXICO     15      6,242      1,302      1      6,059        491   
8         CHICAGO      1      3,078        781      2      1,586        472   
9          DALLAS      4      3,085      1,347     10      5,651        521   

    2021 Unnamed: 9 Unnamed: 10 * 2022 Unnamed: 12 Unnamed: 13  
0  DESC.     HOMBRE       MUJER  DESC.      HOMBRE       MUJER  
1      1         47          13    NaN          11           6  
2    NaN        146          78    NaN          75          55  
3    NaN         91          75    NaN         129          98  
4    NaN         16          25    NaN          38          17  
5    NaN         15           4    NaN          11           3  
6      1      1,279         152      2       1,936         202  
7      2      2,026         262    NaN         409          98  
8      2      1,111         601      4         430         142  
9      1      2,934         277    NaN       1,095          41

Ahora sí podemos crear la función para poder combinar ambas filas en 1. 

In [131]:
def arregla_columnas(dataset_a_modificar):
  encabezados = []
  columna = ""

  encabezados.append(dataset_a_modificar.iloc[0][0])

  for i in range(1, dataset_a_modificar.shape[1], 3):
    columna = dataset_a_modificar.columns[i]
    
    #Para eliminar el asterisco
    if "* " in columna:
      columna = columna.replace("* ", "")

    #Para desc
    encabezados.append('d_' + columna)
    #Para hombre
    encabezados.append('h_' + columna)
    #Para mujer
    encabezados.append('m_' + columna)

  return encabezados

Ya con la función realizamos el ajuste de los encabezados en los 4 datasets.

In [132]:
#Para el dataset de casos
casos.columns = arregla_columnas(casos)
casos.drop(index=0, inplace=True)
casos.head(10)

RUBRO DE PROTECCIÓN d_2010  h_2010 m_2010  \
1     INCIDENTES, MÚLTIPLES INVOLUCRADOS (MIGRATORIO)    NaN     NaN    NaN   
2                      MENOR PROCESADO BAJO LEY TVPRA      1     373    135   
3   ORIENTACIóN GENERAL SOBRE CONSULTAS MIGRATORIA...      2   1,734  1,472   
4                   OTROS CASOS DEL áMBITO MIGRATORIO     10   4,423  1,026   
5    PERMISOS DE INTERNACIóN POR RAZONES HUMANITARIAS      6   1,143  1,352   
6         PROCESOS JUDICIALES MIGRATORIOS DE REMOCIóN     16  10,455  1,174   
7       PROCESOS JUDICIALES MIGRATORIOS DE RESIDENCIA    NaN     339    413   
8             REMOCIóN EXPEDITA EN PUERTOS DE ENTRADA     41  21,203  3,897   
9    REPATRIACIóN DE ENFERMOS EN PROCESOS DE REMOCIóN    NaN     178     25   
10  REPATRIACIóN DE ENFERMOS, PERSONAS ACCIDENTADA...      1     727    327   

   d_2011  h_2011 m_2011 d_2012  h_2012 m_2012  ... m_2019 d_2020  h_2020  \
1     NaN     NaN    NaN    NaN     NaN    NaN  ...    NaN    NaN     NaN   
2     NaN     375    112    NaN     357    194  ...    312      1     230   
3       6   1,844  1,885      4   1,687  1,789  ...  7,948      7   2,584   
4       6   8,073  1,129     11   8,545  6,359  ...  6,519    142  20,194   
5     NaN   1,096  1,403      1     902  1,084  ...  1,045    NaN     510   
6      18  18,514  1,413     26  49,479  1,172  ...  5,339     64  48,940   
7       1     775    304      1     260    212  ...    325      4     580   
8       2  32,763  1,428      2  32,491     94  ...    695      5   2,136   
9      53     158     14    NaN     156     23  ...      8      1     171   
10      1     643    297    NaN     797    331  ...    122      1     331   

   m_2020 d_2021  h_2021 m_2021 d_2022 h_2022 m_2022  
1     NaN    NaN     NaN    NaN    NaN     23      1  
2     128      2     726    526      2    467    359  
3   2,451      4   2,229  1,115      2    693    852  
4   3,248      3  12,009  2,985      1  8,004  1,730  
5     556    NaN     445    476    NaN    126    199  
6   4,564     12  14,908  1,560      9  8,722  1,006  
7     231    NaN     734    296    NaN  1,028     83  
8     535      2   2,286    825      3    564    258  
9      10    NaN      25      2    NaN     12      2  
10    101      3     551    156    NaN    144     55  

[10 rows x 40 columns]

In [133]:
#Para el dataset de oficinas
oficinas.columns = arregla_columnas(oficinas)
oficinas.drop(index=0, inplace=True)
oficinas.head(10)

REPRESENTACIÓN d_2019 h_2019 m_2019 d_2020 h_2020 m_2020 d_2021 h_2021  \
1     ALBUQUERQUE      1    260    173      2    162     59      1     47   
2         ATLANTA      0    660    285      2    287    164    NaN    146   
3          AUSTIN      1    690    429      1    171    144    NaN     91   
4           BOISE      0     82     63      3     62     47    NaN     16   
5          BOSTON      0     61     16      0     16     12    NaN     15   
6     BROWNSVILLE      1  4,899    484      2  4,624    323      1  1,279   
7        CALEXICO     15  6,242  1,302      1  6,059    491      2  2,026   
8         CHICAGO      1  3,078    781      2  1,586    472      2  1,111   
9          DALLAS      4  3,085  1,347     10  5,651    521      1  2,934   
10        DEL RIO      3  1,833    130      0    504     40    NaN    281   

   m_2021 d_2022 h_2022 m_2022  
1      13    NaN     11      6  
2      78    NaN     75     55  
3      75    NaN    129     98  
4      25    NaN     38     17  
5       4    NaN     11      3  
6     152      2  1,936    202  
7     262    NaN    409     98  
8     601      4    430    142  
9     277    NaN  1,095     41  
10     48    NaN    224     50

In [134]:
#Para el dataset de estado de orígen
edo_origen.columns = arregla_columnas(edo_origen)
edo_origen.drop(index=0, inplace=True)
edo_origen.head(10)

EDO. ORIGEN d_2019 h_2019 m_2019 d_2020 h_2020 m_2020 d_2021  \
1        AGUASCALIENTES      1    815    184      0    785     89    NaN   
2       BAJA CALIFORNIA      1    894    316      0    943    215      1   
3   BAJA CALIFORNIA SUR      0     41     23      0     31      8    NaN   
4              CAMPECHE      1    255     46      1    266     21    NaN   
5               CHIAPAS     67  4,437  1,159     21  4,413    673      2   
6             CHIHUAHUA      9  2,353    875      0  1,549    343      2   
7      CIUDAD DE MÉXICO     16  2,519  1,072      6  1,762    567    NaN   
8              COAHUILA      3  1,137    574      0    566    157    NaN   
9                COLIMA      5    303    115      2    261     50    NaN   
10              DURANGO      8  1,292    404      1    826    185    NaN   

   h_2021 m_2021 d_2022 h_2022 m_2022  
1     429     65    NaN    228     28  
2     610    155      1    298     80  
3      22      9    NaN     20     11  
4     162     17    NaN     66     20  
5   3,649    551      3  2,498    445  
6   1,459    279    NaN    835    131  
7     909    438      1    504    238  
8     500    109    NaN    328     56  
9     136     43    NaN     63     24  
10    569    127    NaN    299     84

In [135]:
#Para el dataset de rango de edad
rango_edad.columns = arregla_columnas(rango_edad)
rango_edad.drop(index=0, inplace=True)
rango_edad.head(10)

Grupo de edad d_2019   h_2019  m_2019 d_2020  h_2020  m_2020 d_2021  \
1     De 00 a 17 años     50    7,375   1,985     15   5,569   1,326      1   
2     De 18 a 25 años     78   20,381   4,514     57  18,793   2,827     11   
3     De 26 a 35 años    143   32,117   5,733     82  27,261   3,425      7   
4     De 36 a 45 años     93   24,630   5,599     40  19,468   2,513      7   
5     De 46 a 55 años     48   10,254   3,492     17   7,845   1,555      2   
6     De 56 a 65 años     12    2,646   1,240      2   1,773     532      1   
7     De 66 a 75 años      9      597     369      0     385     197      0   
8      Más de 75 años      5      191     188      0      91      89      0   
9   Se desconoce Edad     15    2,146     595     31   1,864     418     10   
10              TOTAL    453  100,337  23,715    244  83,049  12,882     39   

    h_2021 m_2021 d_2022  h_2022 m_2022  
1   11,737  2,128      8   7,288  1,325  
2   10,538  1,903      5   4,541    995  
3   11,122  2,021      2   6,208  1,154  
4    8,729  1,570      5   4,986    828  
5    4,143  1,052      4   2,125    588  
6    1,044    400      1     684    222  
7      225    150      0     119     87  
8       76     81      0      32     36  
9    1,494    376      2   1,875    325  
10  49,108  9,681     27  27,858  5,560

Proseguimos ahora a revisar los NaNs de nuestros 4 datasets. Los contamos sumando (.sum) todos los valores True que arroje la función ".isna()". Esto funciona, ya que pandas reconoce True como 1 y False como 0. Por lo tanto obtendremso cuantos NaNs hay en nuestros datasets. 

Primero por columna:

In [136]:
#NaNs por columna - casos
casos.isna().sum()

RUBRO DE PROTECCIÓN     0
d_2010                  5
h_2010                  1
m_2010                  1
d_2011                  6
h_2011                  1
m_2011                  1
d_2012                  7
h_2012                  1
m_2012                  1
d_2013                  9
h_2013                  1
m_2013                  1
d_2014                  8
h_2014                  1
m_2014                  1
d_2015                  5
h_2015                  1
m_2015                  1
d_2016                  7
h_2016                  1
m_2016                  1
d_2017                  5
h_2017                  1
m_2017                  1
d_2018                 13
h_2018                  1
m_2018                  1
d_2019                  4
h_2019                  1
m_2019                  1
d_2020                  5
h_2020                  1
m_2020                  1
d_2021                  8
h_2021                  1
m_2021                  1
d_2022                  9
h_2022      

In [137]:
#NaNs por columna - oficinas
oficinas.isna().sum()

REPRESENTACIÓN     0
d_2019             0
h_2019             0
m_2019             0
d_2020             0
h_2020             0
m_2020             0
d_2021            28
h_2021             0
m_2021             0
d_2022            38
h_2022             0
m_2022             0
dtype: int64

In [138]:
#NaNs por columna - estado de orígen
edo_origen.isna().sum()

EDO. ORIGEN    16
d_2019         16
h_2019         16
m_2019         16
d_2020         16
h_2020         16
m_2020         16
d_2021         37
h_2021         16
m_2021         16
d_2022         37
h_2022         16
m_2022         16
dtype: int64

In [139]:
#NaNs por columna - rango edad
rango_edad.isna().sum()

Grupo de edad    41
d_2019           41
h_2019           41
m_2019           41
d_2020           41
h_2020           41
m_2020           41
d_2021           41
h_2021           41
m_2021           41
d_2022           41
h_2022           41
m_2022           41
dtype: int64

Y después por fila:

In [140]:
#NaNs por fila - casos
casos.isna().sum(axis = 1)

1     37
2      5
3      0
4      0
5      5
6      0
7      7
8      1
9     11
10     7
11     1
12     9
13     4
14     5
15    12
16    12
dtype: int64

In [141]:
#NaNs por fila - oficinas
oficinas.isna().sum(axis = 1)

1     1
2     2
3     2
4     2
5     2
6     0
7     1
8     0
9     1
10    2
11    1
12    2
13    2
14    0
15    0
16    1
17    1
18    2
19    2
20    2
21    0
22    2
23    2
24    2
25    1
26    2
27    2
28    0
29    2
30    2
31    2
32    2
33    2
34    1
35    2
36    2
37    0
38    1
39    2
40    0
41    1
42    1
43    1
44    2
45    1
46    1
47    2
48    0
49    1
50    1
51    0
dtype: int64

In [142]:
#NaNs por fila - edo_origen
edo_origen.isna().sum(axis = 1)

1      2
2      0
3      2
4      2
5      0
6      1
7      1
8      2
9      2
10     2
11     0
12     0
13     0
14     0
15     0
16     0
17     2
18     2
19     2
20     2
21     0
22     1
23     2
24     2
25     1
26     2
27     2
28     2
29     1
30     2
31     2
32     2
33     1
34     0
35     0
36    13
37    13
38    13
39    13
40    13
41    13
42    13
43    13
44    13
45    13
46    13
47    13
48    13
49    13
50    13
51    13
dtype: int64

In [143]:
#NaNs por fila - rango_edad
rango_edad.isna().sum(axis = 1)

1      0
2      0
3      0
4      0
5      0
6      0
7      0
8      0
9      0
10     0
11    13
12    13
13    13
14    13
15    13
16    13
17    13
18    13
19    13
20    13
21    13
22    13
23    13
24    13
25    13
26    13
27    13
28    13
29    13
30    13
31    13
32    13
33    13
34    13
35    13
36    13
37    13
38    13
39    13
40    13
41    13
42    13
43    13
44    13
45    13
46    13
47    13
48    13
49    13
50    13
51    13
dtype: int64

In [ ]:
#df_dropped = df.dropna(axis=0, how='all')
#df_no_nans['productos_vendidos'].fillna(0)
#df_dropped = df_dropped.reset_index(drop=True)
#column_name_mapping = {
#    'Suburb': 'suburb',
#    'Address': 'address',
#}
#df_renamed = df_dropped.rename(columns=column_name_mapping)

##F. Transformación de datos  - Postwork 7


In [ ]:
#diccionario_de_conversion = {
#    'bestsellers_date.numberLong': 'datetime64[ns]',
#    'published_date.numberLong': 'datetime64[ns]'
#}
#temp = df.astype(diccionario_de_conversion)
#pd.to_datetime(df['published_date.numberLong'], unit='ms')
#df['rank.numberInt'] = pd.to_numeric(df['rank.numberInt'], errors='coerce')
#df['description'] = df['description'].str.replace('Descr:', '')
#df['description'] = df['description'].str.strip()
#df['title'].str.lower()
#df[['author_first_name', 'author_last_name']] = df['author'].str.split(' ', expand=True)
#def years_since_bestseller(value):
#    as_datetime = pd.to_datetime(value, unit='ms')
#    today = pd.to_datetime('today')
#    difference_in_days = (today - as_datetime).days
#    in_years = difference_in_days / 365    
#    return in_years
#df['published_date.numberLong'].apply(years_since_bestseller)


##G. Planes a futuro  - Postwork 8

Dado que este módulo únicamente abarca el procesamiento de datos, y por ende no se finaliza la investigación, proponemos los siguientes puntos para un futuro estudio del tema:

*   Obtener datos más concretos respecto al tipo de trámites realizados en cada oficina del consulado.
*   Comparar los datos de este dataset con las estadísticas generales de migración para...

##H. Referencias

1.   https://www.cndh.org.mx/derechos-humanos/derechos-de-las-personas-migrantes
2.   https://consulmex.sre.gob.mx/washington/index.php/prot
3.   https://datos.gob.mx/busca/dataset/migratorio--casos-de-proteccion-y-o-asistencia-consular-atendidos-por-la-rdcm-en-el-mundo/resource/3b99534a-6bf6-4f43-91f2-e5281e017473
4.   https://firstfocus.org/wp-content/uploads/2014/08/Legal-Protections-for-Unaccompanied-Minors-in-the-Trafficking-Victims-Protection-Act-of-2008.pdf
5.   https://www.cndh.org.mx/sites/all/doc/cartillas/2015-2016/37-Migrantes-DH.pdf